casper/src/main/resources/MultiSigRevVault.rho

```diff
@@ -0,0 +1,212 @@
+                                      ret!((true, (false, "done")))
+                                    } else {
+                                      ret!((false, msg))
+                                    }
+                                  }
+                                }
+                              } else {
+                                @(*multiSig, *nonceMapStore)!(
+                                  nonceMap.set(nonce, (targetRevAddress, amount, Set(authKey)))
+                                ) |
+                                ret!((true, (true, nonce)))
+                              }
+                            }
+                          } else {
+                            ret!((false, "invalid authKey"))
+                          }
+                        }
+                      }
+                    } |
+
+                    // Confirms transfer transaction.
+                    // returns (true, (true, nonce)) when more sigs required
+                    //         (true, (false, "done")) when only one sig required
+                    //         (false, errorMsg) on failure
+                    contract multiSig(@"confirm", @targetRevAddress, @amount, @authKey, @nonce, ret) = {
+                      // Check that authKeys is in the list of authKeys or authPubKeys
+                      new idxAuthKeysCh, idxAuthPubKeysCh in {
+                        @ListOps!("indexOf", authKeys, authKey, *idxAuthKeysCh) |
+                        @ListOps!("indexOf", authPubKeys, authKey, *idxAuthPubKeysCh) |
+                        for (@idxAuthKeys <- idxAuthKeysCh; @idxAuthPubKeys <- idxAuthPubKeysCh) {
+                          if (idxAuthKeys >= 0 or idxAuthPubKeys >= 0) {
+                            for (@nonceMap <- @(*multiSig, *nonceMapStore)) {
+                              match nonceMap.get(nonce) {
+                                Nil => {
+                                  @(*multiSig, *nonceMapStore)!(nonceMap) |
+                                  ret!((false, "inactive transaction nonce"))
+                                }
+                                (nonceTargetRevAddress, nonceAmount, nonceConfirmers) => {
+                                  if (targetRevAddress != nonceTargetRevAddress or
+                                      amount != nonceAmount) {
+                                    @(*multiSig, *nonceMapStore)!(nonceMap) |
+                                    ret!((false, "given data does not match stored data"))
+                                  } else {
+                                    if (nonceConfirmers.contains(authKey)) {
+                                      @(*multiSig, *nonceMapStore)!(nonceMap) |
+                                      ret!((false, "already confirmed"))
+                                    } else {
+                                      // If this is the last confirmation
+                                      if (nonceConfirmers.size() == quorumSize - 1) {
```

Comment:
What if two valid confirms get executed in parallel when we are in the state of `nonceConfirmers.size() == quorumSize - 1`? Are you sure that there is no chance that the RevVault transfer won't be executed twice?

Reply:
The two confirm calls would race for acquisition of the nonceMap, which behaves like a mutex.  One would acquire it first and update the set of confirmers to Nil.  The other would acquire it second and would receive the "inactive transaction nonce" message.

Reply:
But currently the following seems possible:
First confirm sets nonceMap
Second confirm can acquire nonceMap
First confirm calls the RevVault transfer
Second confirm calls the RevVault transfer again

I think what needs to be done is to set the nonceMap after the RevVault transfer executed.